File: solution2.py
# Time: O(n)
# space: O(n)
class Solution(object):
    def rob(self, nums):
        if len(nums) == 1:
            return nums[0]
        dp = [0]*(len(nums))
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])
        for i in range(2, len(nums)):
            dp[i] = max(dp[i-1], nums[i] + dp[i-2])
        return dp[len(nums) - 1]

File: Solution2.java
public class Solution2 {
    public int rob(int[] nums) {
        int[] total = new int[nums.length + 1];
        total[0] = 0;
        total[1] = nums[0];
        for(int i = 2; i < total.length; i++){
            total[i] = Math.max(total[i-1], total[i-2] + nums[i-1]);
        }
        return total[nums.length];
    }
}
